org.json.JSONException: No value for properties

```
E/AndroidRuntime( 2489): Caused by: org.json.JSONException: No value for properties
E/AndroidRuntime( 2489): 	at org.json.JSONObject.get(JSONObject.java:389)
E/AndroidRuntime( 2489): 	at org.json.JSONObject.getJSONObject(JSONObject.java:609)
E/AndroidRuntime( 2489): 	at com.schedjoules.client.eventsdiscovery.json.JsonLink.parameters(JsonLink.java:127)
E/AndroidRuntime( 2489): 	at com.schedjoules.client.eventsdiscovery.json.JsonLink.firstParameter(JsonLink.java:119)
E/AndroidRuntime( 2489): 	at com.schedjoules.eventdiscovery.model.ParcelableLink.firstParameter(ParcelableLink.java:119)
E/AndroidRuntime( 2489): 	at com.schedjoules.eventdiscovery.model.ParcelableLink.writeToParcel(ParcelableLink.java:153)
E/AndroidRuntime( 2489): 	at android.os.Parcel.writeParcelable(Parcel.java:1357)
E/AndroidRuntime( 2489): 	at android.os.Parcel.writeValue(Parcel.java:1262)
E/AndroidRuntime( 2489): 	at android.os.Parcel.writeList(Parcel.java:711)
E/AndroidRuntime( 2489): 	at android.os.Parcel.writeValue(Parcel.java:1284)
E/AndroidRuntime( 2489): 	at android.os.Parcel.writeArrayMapInternal(Parcel.java:638)
```

--- a/eventsdiscovery/src/main/java/com/schedjoules/client/eventsdiscovery/json/JsonLink.java
+++ b/eventsdiscovery/src/main/java/com/schedjoules/client/eventsdiscovery/json/JsonLink.java
@@ -124,7 +124,7 @@ public final class JsonLink implements Link
     @Override
     public <T> Iterator<Parameter<T>> parameters(final ParameterType<T> parameterType)
     {
-        final JSONObject properties = mJsonObject.getJSONObject(KEY_PROPERTIES);
+        final JSONObject properties = mJsonObject.optJSONObject(KEY_PROPERTIES);
         return properties == null ? EmptyIterator.<Parameter<T>>instance() : new ConvertedIterator<>(
                 new FilteredIterator<>(properties.keys(), new AnyOf<>(parameterType.name())),
                 new AbstractConvertedIterator.Converter<Parameter<T>, String>()
@@ -141,7 +141,7 @@ public final class JsonLink implements Link
     @Override
     public <T> boolean hasParameter(ParameterType<T> parameterType)
     {
-        JSONObject properties = mJsonObject.getJSONObject(KEY_PROPERTIES);
+        JSONObject properties = mJsonObject.optJSONObject(KEY_PROPERTIES);
         return properties != null && properties.has(parameterType.name());
     }
 }

--- a/eventsdiscovery/src/test/java/com/schedjoules/client/eventsdiscovery/json/JsonLinkTest.java
+++ b/eventsdiscovery/src/test/java/com/schedjoules/client/eventsdiscovery/json/JsonLinkTest.java
@@ -1,0 +1,179 @@
+/*
+ * Copyright 2016 SchedJoules
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.schedjoules.client.eventsdiscovery.json;
+
+import org.dmfs.httpessentials.converters.PlainStringHeaderConverter;
+import org.dmfs.httpessentials.parameters.BasicParameterType;
+import org.dmfs.httpessentials.parameters.ParameterType;
+import org.dmfs.httpessentials.types.StructuredMediaType;
+import org.json.JSONObject;
+import org.junit.Test;
+
+import java.net.URI;
+import java.util.Arrays;
+import java.util.HashSet;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+
+/**
+ * @author Marten Gajda
+ */
+public class JsonLinkTest
+{
+    @Test
+    public void target() throws Exception
+    {
+        assertEquals(URI.create("http://example.com"), new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\"\n" +
+                "}")).target());
+    }
+
+
+    @Test
+    public void context() throws Exception
+    {
+        assertEquals(URI.create("http://test.com"), new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\"\n" +
+                "}")).context(URI.create("http://test.com")));
+    }
+
+
+    @Test
+    public void languages() throws Exception
+    {
+        // TODO: implement test
+    }
+
+
+    @Test
+    public void title() throws Exception
+    {
+        assertEquals("link-title", new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"link-title\",\n" +
+                "  \"href\": \"http://example.com\"\n" +
+                "}")).title());
+    }
+
+
+    @Test
+    public void mediaType() throws Exception
+    {
+        assertEquals(new StructuredMediaType("application", "xml"), new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"link-title\",\n" +
+                "  \"href\": \"http://example.com\",\n" +
+                "  \"type\": \"application/xml\"\n" +
+                "}")).mediaType());
+    }
+
+
+    @Test
+    public void relationTypes() throws Exception
+    {
+        assertEquals(new HashSet<>(Arrays.asList("rel-type")), new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\"\n" +
+                "}")).relationTypes());
+    }
+
+
+    @Test
+    public void reverseRelationTypes() throws Exception
+    {
+        assertEquals(new HashSet<>(), new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\"\n" +
+                "}")).reverseRelationTypes());
+    }
+
+
+    @Test
+    public void firstParameter() throws Exception
+    {
+        ParameterType<String> parameterType = new BasicParameterType<>("test", new PlainStringHeaderConverter());
+        assertEquals("123", new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\",\n" +
+                "  \"properties\": {\n" +
+                "     \"test\": \"123\"\n" +
+                "  }\n" +
+                "}")).firstParameter(parameterType, "fail").value());
+    }
+
+
+    @Test
+    public void firstParameter2() throws Exception
+    {
+        ParameterType<String> parameterType = new BasicParameterType<>("test", new PlainStringHeaderConverter());
+        assertEquals("123", new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\",\n" +
+                "  \"properties\": {\n" +
+                "  }\n" +
+                "}")).firstParameter(parameterType, "123").value());
+    }
+
+
+    @Test
+    public void parameters() throws Exception
+    {
+        // TODO: implement test
+    }
+
+
+    @Test
+    public void hasParameter() throws Exception
+    {
+        ParameterType<String> parameterType = new BasicParameterType<>("test", new PlainStringHeaderConverter());
+        assertTrue(new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\",\n" +
+                "  \"properties\": {\n" +
+                "    \"test\": \"123\"\n" +
+                "  }\n" +
+                "}")).hasParameter(parameterType));
+    }
+
+
+    @Test
+    public void hasParameter2() throws Exception
+    {
+        ParameterType<String> parameterType = new BasicParameterType<>("test", new PlainStringHeaderConverter());
+        assertFalse(new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\",\n" +
+                "  \"properties\": {\n" +
+                "  }\n" +
+                "}")).hasParameter(parameterType));
+    }
+
+}

--- a/eventsdiscovery/src/test/java/com/schedjoules/client/eventsdiscovery/json/JsonLinkTest.java
+++ b/eventsdiscovery/src/test/java/com/schedjoules/client/eventsdiscovery/json/JsonLinkTest.java
@@ -140,6 +140,16 @@ public class JsonLinkTest
                 "}")).firstParameter(parameterType, "123").value());
     }
 
+    @Test
+    public void firstParameter3() throws Exception
+    {
+        ParameterType<String> parameterType = new BasicParameterType<>("test", new PlainStringHeaderConverter());
+        assertEquals("123", new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\"\n" +
+                "}")).firstParameter(parameterType, "123").value());
+    }
 
     @Test
     public void parameters() throws Exception
@@ -176,4 +186,15 @@ public class JsonLinkTest
                 "}")).hasParameter(parameterType));
     }
 
+
+    @Test
+    public void hasParameter3() throws Exception
+    {
+        ParameterType<String> parameterType = new BasicParameterType<>("test", new PlainStringHeaderConverter());
+        assertFalse(new JsonLink(new JSONObject("{\n" +
+                "  \"rel\": \"rel-type\",\n" +
+                "  \"title\": \"title\",\n" +
+                "  \"href\": \"http://example.com\"\n" +
+                "}")).hasParameter(parameterType));
+    }
 }

